### PR TITLE
Week 4

### PR DESCRIPTION
**Pull request by Nabraha**

### CODE REVIEW

File: week-3/A-array-find/exercise.js [hunk 1]
```diff
@@ -5,7 +5,9 @@
 */
 
 // -- Complete this function -->
-function findLongNameThatStartsWithA(names) {}
+function findLongNameThatStartsWithA(names) {
```
Comment (p-maxwell) [inline week-3/A-array-find/exercise.js:8]: This is good and correct. You need to be a little careful of merge conflicts with doing week-3 stuff in the week-4 pull request though.

File: week-4/C-more-complex-objects/exercise-3.js [hunk 2]
```diff
@@ -31,13 +31,33 @@ var parkAvenueHouse = {
...
+  return house.currentOwner.firstName+ ' ' + house.currentOwner.lastName;
+}
 
 // returns an array of the owners' email addresses of the two houses
-function getEmailAddresses(house1, house2) {}
+function getEmailAddresses(house1, house2) {
+  var arr = [];
+  //arr[0] = house1.currentOwner.email; 
+  //arr[1] = house2.currentOwner.email;
+  arr.push(house1.currentOwner.email);//at the end of the array
```
Comment (p-maxwell) [inline week-4/C-more-complex-objects/exercise-3.js:43]: Nice way of doing it. Everything before this in week 4 has been good by the way - you've aced it so far, so I've had no comments.

File: week-4/C-more-complex-objects/exercise-3.js [hunk 3]
```diff
@@ -31,13 +31,33 @@ var parkAvenueHouse = {
...
+function getCheapestAddress(house1, house2) {
+  var houseOnePrice = house1.price;
+  var houseTwoPrice = house2.price;
+
+  if(houseOnePrice < houseTwoPrice){
+    return house1.address;
+  } else if (houseOnePrice > houseTwoPrice){
+    return house2.address;
+  }else{
+    return 'price are the same';
```
Comment (p-maxwell) [inline week-4/C-more-complex-objects/exercise-3.js:58]: Be careful here. If the caller of getCheapestAddress expects to get a house address back and they get this text instead, it might be a bit of a shock if they try to do anything with it (for instance, look up where it is in a list).

It's safe to return either address here, both are equally cheap.

File: week-4/D-methods/exercise-5.js [hunk 4]
```diff
@@ -16,8 +16,22 @@ var coffeeMachine = {
...
-  getCoffee: function(coffee) {}
+
+ // insertMoney takes an amount in parameter to add money in the coffee machine.
+  insertMoney: function(amount) {
+    this.insertedAmount = amount;
+  },
+
+  //getCoffee takes a coffee type in parameter and dispends the selected coffee
+  //only if the inserted amount is greater or equal than the price of the coffee!
+  getCoffee: function(coffee) {
```
Comment (p-maxwell) [inline week-4/D-methods/exercise-5.js:27]: I remember this one. :)

File: week-4/E-arrays-of-objects/exercise-1.js [hunk 5]
```diff
@@ -24,12 +24,25 @@ DO NOT EDIT ANYTHING ABOVE THIS LINE
...
+
+//2) Return an array of the person names(hint: use map).
+var personNames =  persons.map(function(element){
+    return element.name;
+});// Complete here
+
+//3) Filter the persons to return an array with the person younger than 28 years old(hint: use filter).
+var personsYoungerThan28YearsOld = persons.filter(function(element){
+    if(element.age < 28){
+        return true;
```
Comment (p-maxwell) [inline week-4/E-arrays-of-objects/exercise-1.js:43]: You can shorten this to `return element.age < 28` or even the whole function to `element => element.age < 28`.

File: week-4/E-arrays-of-objects/exercise-1.js [hunk 6]
```diff
@@ -24,12 +24,25 @@ DO NOT EDIT ANYTHING ABOVE THIS LINE
...
+    } 
+    //return element.age < 28 ? true : false // Elvis
```
Comment (p-maxwell) [inline week-4/E-arrays-of-objects/exercise-1.js:45]: Important note: this code does not run for me. You need }); somewhere about here to finish the function and call to filter().

File: week-4/E-arrays-of-objects/exercise-2.js [hunk 7]
```diff
@@ -39,12 +39,30 @@ DO NOT EDIT ANYTHING ABOVE THIS LINE
...
+  return element.destinationName;
+})// Complete here
 
-var destinationNamesMoreThan300KmsAwayByTrain = // Complete here (PRINT THE RESULT IN THE CONSOLE USING FOREACH)
+//3) Print in the console all the destination names more than 300 kms far away and reachable by train.
+var destinationNamesMoreThan300KmsAwayByTrain = travelDestinations.filter(function(element1){
+  return element1.distanceKms > 300 && element1.transportations.includes("train");
+}).map(function(element){
+  return element.destinationName;
+})// Complete here (PRINT THE RESULT IN THE CONSOLE USING FOREACH)
```
Comment (p-maxwell) [inline week-4/E-arrays-of-objects/exercise-2.js:65]: Nice. Some trickier stuff in here. All correct.

File: week-4/E-arrays-of-objects/exercise-3.js [hunk 8]
```diff
@@ -58,14 +58,34 @@ var restaurantFinderApplication = {
...
+      }
+    )
   },
+
+/*3) Define a method countNumberOfRestaurantsInArea which takes an area of Glasgow in parameter (centre, west),
+and returns the number of restaurants in this area.
+*/
   countNumberOfRestaurantsInArea: function(area) {
-    // Complete here
+    return restaurants.filter(element => element.address.area.includes(area)).length
```
Comment (p-maxwell) [inline week-4/E-arrays-of-objects/exercise-3.js:88]: All correct again.

File: week-4/F-object-keys/exercise-part-2.js [hunk 9]
```diff
@@ -33,14 +33,14 @@ var storeBranches = {
...
 
 // # 2
 // prints [ 'manager', 'assistant', 'interns' ]
-console.log();
+console.log(Object.keys(storeBranches.glasgow));
 
 // # 3
 // prints [ 'head_intern', 'intern' ]
-console.log();
+console.log(Object.keys(storeBranches.edinburgh.interns));
```
Comment (p-maxwell) [inline week-4/F-object-keys/exercise-part-2.js:44]: And all the Object keys stuff is correct. Looks like the week-4 stuff didn't pose you too many problems.

File: week-4/E-arrays-of-objects/exercise-1.js (hunk 6 above)
Comment (Nabraha) [reply week-4/E-arrays-of-objects/exercise-1.js:45]: Will commit the change

File: week-4/E-arrays-of-objects/exercise-1.js (hunk 5 above)
Comment (Nabraha) [reply week-4/E-arrays-of-objects/exercise-1.js:43]: I will xommot the change
